storage/tests/test_mvcc.rs

```diff
@@ -34,3 +34,137 @@ This file should comprise a set of low-level tests relating to MVCC.
+                match *self { $(Self::$variant => $name),* }
+            }
+        }
+    };
+}
+
+test_keyspace_set! {
+    Keyspace => 0: "keyspace",
+}
+
+use self::TestKeyspaceSet::Keyspace;
+
+const KEY_1: [u8; 4] = [0x0, 0x0, 0x0, 0x1];
+const KEY_2: [u8; 4] = [0x0, 0x0, 0x0, 0x2];
+const VALUE_0: [u8; 1] = [0x0];
+const VALUE_1: [u8; 1] = [0x1];
+const VALUE_2: [u8; 1] = [0x2];
+
+fn setup_storage(storage_path: &Path) -> MVCCStorage<WAL> {
+    let storage = MVCCStorage::recover::<TestKeyspaceSet>("storage", storage_path).unwrap();
+    storage
+}
+
+#[test]
+fn test_commit_increments_watermark() {
+    init_logging();
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+    let wm_initial = storage.read_watermark();
+    let snapshot_0 = storage.open_snapshot_write();
+    snapshot_0.commit().unwrap();
+
+    assert_eq!(wm_initial.next().number().number() + 1, storage.read_watermark().number().number());
+}
+
+#[test]
+fn test_reading_snapshots() {
+    init_logging();
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let key_1: &StorageKey<'_, 48> = &StorageKey::Reference(StorageKeyReference::new(Keyspace, ByteReference::new(&KEY_1)));
+
+    let snapshot_write_0 = storage.open_snapshot_write();
+    snapshot_write_0.put_val(StorageKeyArray::new(Keyspace, ByteArray::copy(&KEY_1)), ByteArray::copy(&VALUE_0)).unwrap();
+    snapshot_write_0.commit().unwrap();
+
+    let watermark_0 = storage.read_watermark();
+
+    let snapshot_read_0 = storage.open_snapshot_read();
+    assert_eq!(snapshot_read_0.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_0);
+
+    let snapshot_write_1 = storage.open_snapshot_write();
+    snapshot_write_1.put_val(StorageKeyArray::new(Keyspace, ByteArray::copy(&KEY_1)), ByteArray::copy(&VALUE_1)).unwrap();
+    let snapshot_read_01 = storage.open_snapshot_read();
+    assert_eq!(snapshot_read_0.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_0);
+    assert_eq!(snapshot_read_01.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_0);
+
+    let result_write_1 = snapshot_write_1.commit();
+    assert!(result_write_1.is_ok());
+
+    let snapshot_read_1 = storage.open_snapshot_read();
+    assert_eq!(snapshot_read_0.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_0);
+    assert_eq!(snapshot_read_01.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_0);
+    assert_eq!(snapshot_read_1.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_1);
+    snapshot_read_1.close_resources();
+    snapshot_read_01.close_resources();
+    snapshot_read_0.close_resources();
+
+    // Read from further in the past.
+    let snapshot_read_02 = storage.open_snapshot_read_at(watermark_0);
+    assert_eq!(snapshot_read_02.get::<128>(key_1.as_reference()).unwrap().unwrap().bytes(), VALUE_0);
+    snapshot_read_02.close_resources();
+}
+
+#[test]
```

Comment:
I'm not sure this test is meaningful in our context given

1. We may not need to fail transactions which concurrently write the same key.
2. We have isolation tests to test isolation is enforced.

Reply:
Hang on, just from reading the title, in our semantics writing the same key twice should not fail right?